### PR TITLE
add support to detect ts-jest environment

### LINKED ISSUE #86: Jest -> ts-jest + Typescript does not appear to be working
## 🐛 Bug Report

In Fastify-Autoload 1.x jest tests + typescript worked with fastify-autoload with seemingly no issue as long as the `includeTypescript: true` configuration was passed.  With the new version it appears that has been removed in favor of automatically supporting ts-node.  I don't 100% know everything about ts-jest but I do not believe it uses ts-node and yet it worked before.  It seems like this should still work.

## To Reproduce

clone this repo - https://github.com/wolfejw86/fastify-typescript-starter

run `npm install`
run `npm run test`

see that fasitify-autoload is working correctly in https://github.com/wolfejw86/fastify-typescript-starter/blob/master/src/app.ts

In the same repo - upgrade fastify-swagger, fastify, fastify-plugin, and fastify-autoload all to the latest version by switching to the branch in progress(the 3.0 migration for fastify + typescript).
`git checkout fastify-upgrade`
`npm install`
`npm run test`

Notice fastify-autoload will no longer load any plugins.

```
fastify-autoload cannot import plugin at '/Users/johnwolfe/Desktop/repositories/fastify-typescript-starter/src/plugins/globalErrorHandler.ts'. To fix this error compile TypeScript to JavaScript or use 'ts-node' to run your app.

      at findPlugins (node_modules/fastify-autoload/index.js:87:15)
      at fastifyAutoload (node_modules/fastify-autoload/index.js:19:19)
```

Edit:
Additionally if I hardcode this line:
image

~It _mostly_ works, but doesn't load an `index.ts` file in the services folder from:~ EDIT: it works fine - I did not correctly understand the new behavior of how autoload plugin handles nested folders
~I am waiting for `fastify.ready()` but it seems like it's firing before the second call to the autoload plugin has finished resolving or something.~ EDIT: It loads all modules correctly according to how Autoload uses nested folders as pathnames now.

Any help is appreciated :)

### PR DESCRIPTION
**Pull request by ferco0**
#### Checklist

Fixes #86 

- [x] run `npm run test` ~~and `npm run benchmark`~~
- [x] tests ~~and/or benchmarks are~~ included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
This is great! Can you add a link/reference where you
 did find this?

**Comment by ferco0** [comment]
Thanks! I dumped the process object and looked for something that could help.

**Comment by mcollina** [comment]
could you add a simple test with jest as well? I don't want this to break without notice on a new release.

File: package.json [hunk 1]
```diff
@@ -9,8 +9,9 @@
     "unit": "if semver `node --version` --range '>=14 || >= 12.17.0 < 13.0.0'; then npm run unit:with-modules; else npm run unit:without-modules; fi",
     "unit:with-modules": "tap test/*/*.js test/*/*.ts",
     "unit:without-modules": "tap test/commonjs/*.js test/*/*.ts",
-    "test": "npm run lint && npm run unit && npm run typescript",
-    "typescript": "tsd"
+    "test": "npm run lint && npm run unit && npm run typescript && npm run typescript:jest",
+    "typescript": "tsd",
+    "typescript:jest": "npx jest"
```
Comment (RafaelGSS) [inline package.json:14]:
```suggestion
    "typescript:jest": "jest"
```
You already installed at devDependencies.

File: test/typescript-jest/basic/app.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { FastifyPlugin } from 'fastify'
+import { join } from 'path'
+const fastifyAutoLoad = require('../../../')
```
Comment (mcollina) [inline test/typescript-jest/basic/app.ts:3]: Could you use import instead?

Comment (ferco0) [reply test/typescript-jest/basic/app.ts:3]: I don't think this is needed, will need install babel-jest and add more settings.

**Comment by wolfejw86** [comment]
fwiw the property referenced on process.env always comes up `undefined` for me.  I'm not sure this works with all versions of ts-jest 🤔 

Is it documented anywhere that it is setting this variable on the process?

**Comment by wolfejw86** [comment]
With further inspection - this flag `process.env.npm_package_jest_transform_______ts_tsx__` seems to _only_ be set when your jest config settings live inside your package.json file.  If this change goes in as is, it may only support repos that use jest configuration that way which might be a more narrow scope than was intended?  I know in my circumstances (and I believe is common) I've always used a configuration file separate from `package.json`

**Comment by mcollina** [comment]
@wolfejw86 wdyt, should this land or not?

**Comment by wolfejw86** [comment]
I absolutely don't want to take away from the work @ferco0 did - it's really cool to take a deep dive and see how running things through an `npm` command enhances the `process.env` object.  Unfortunately this solution does not solve the referenced issue when running in a repository via a separate jest config file _nor_ does it work when running without the use of an npm command (common in repos that use `yarn` or repos that deploy via node-slim containers where you do `node server.js` to start the app).
Therefore my 2 cents is that since it doesn't fully solve #82 I would not recommend it to land.

**Comment by ferco0** [comment]
I installed yarn to test it, I put the jest settings in `jest.config.js` and it works for me:



I think that don't need check the Typescript support if executes a script with node command.

**Comment by wolfejw86** [comment]
That’s fair about the node command part. I didn’t realize it worked for
yarn too. The bot working with a heat config file still gets me though

**Comment by wolfejw86** [comment]
Hmmm, I definitely can't reproduce it anymore.  I was probably doing something silly in tandem with testing this out.  It appears to work fine.  With the assumption that you aren't ever going to need to run this with a barebones nodejs command I think it's good to go.  Sorry for the hold up 🤦

**Comment by ferco0** [comment]
@wolfejw86 thanks man, you helped improve the detection! :pray:

**Comment by johanneslumpe** [comment]
@mcollina For what it's worth, this fix does not work for me. `process.env.npm_package_jest_transform_______ts_tsx__` is not defined when I run my tests with a custom config file. Is there a chance we can get the `includeTypeScript` flag back instead?

**Comment by mcollina** [comment]
a PR is always welcomed.

**Comment by johanneslumpe** [comment]
Turns out I cannot reproduce it anymore. Not idea what happened, sorry for the confusion.

**Pull request merged**